Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +124,296 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [← toDFA_correct] at hx ⊢
   exact M.toDFA.pumping_lemma hx hlen
 
+/-! ### Regex-like operations
+
+Since NFAs are models for regular languages, they provide decision procedures for the empty set,
+singleton sets with members of at most length 1 as well as a notion of closedness under union,
+concatenation and application of the Kleene star.
+
+Cf. <https://en.wikipedia.org/wiki/Thompson%27s_construction> for the basic idea in a setting
+permitting ε-moves.
+-/
+instance : Zero (NFA α σ) :=
+  ⟨default⟩
+
+@[simp]
+theorem zero_correct : (0 : NFA α σ).accepts = 0 := by
+  ext
+  refine ⟨?_, False.elim⟩
+  rintro ⟨_, ⟨⟩, _⟩
+
+instance : One (NFA α σ) :=
+  ⟨⟨fun _ _ ↦ ∅, univ, univ⟩⟩
+
+@[simp]
+theorem one_correct [Inhabited σ] :
+    (1 : NFA α σ).accepts = 1 := by
+  ext x
+  constructor
+  · rintro ⟨q, _, eval⟩
+    rcases x.eq_nil_or_concat with rfl | ⟨_, _, rfl⟩
+    · exact rfl
+    rw [List.concat_eq_append, eval_append_singleton, mem_stepSet] at eval
+    rcases eval with ⟨_, _, ⟨⟩⟩
+  · rintro rfl
+    exact ⟨default, Set.mem_univ _, Set.mem_univ _⟩
+
+/-- An NFA accepting just a single word of length 1, given by the underlying character. -/
+def char (a : α) : (NFA α (σ ⊕ σ')) :=
+  ⟨fun p c q ↦ p.isLeft ∧ a = c ∧ q.isRight, (Sum.isLeft ·), (Sum.isRight ·)⟩
```

Comment:
```suggestion
/-- An NFA accepting just a single word of length 1, given by the underlying character. -/
def char (a : α) : NFA α (σ ⊕ σ') :=
  ⟨fun p c q ↦ p.isLeft ∧ a = c ∧ q.isRight, (Sum.isLeft ·), (Sum.isRight ·)⟩
```

Reply:
Also, I don't really understand why we need `σ` and `σ'` around here. Why not just `Fin 2` as the set of states?

Reply:
It is a bit awkward, to be sure. I wanted to stay as generic as possible and what really matters about the `char` automaton's state type is that it has at least two distinct elements. Unfortunately, there doesn't seem to be a data-carrying pendant to `Nontrivial`, like `Inhabited` is to `Nonempty`. There is `TwoPointed`, but that's not a class…

`Fin 2` (or `Bool` or whatever) would lock the state space in place.
This way, one can't just conjure up a three-stated `char` automaton and re-use the `char_correct` proof (if that is ever on someone's mind).
It also raises the question of why the `One` instance doesn't just work with `Fin 1`/`Unit` or `Zero` with `Fin 0`/`Empty`.

But in the end, I'm really curious to find out what “the mathlib approach”™ would be. I see four different ways to proceed here:
1. make `TwoPointed` a class (adding another PR?)
2. use an `Inhabited σ` and let `char` work with `Option σ`
3. keep the sum
4. use a specific state type (and have a think about `Zero`/`One`?)

I would be glad to receive some input and/or additions!